Lab0 - Tour of rust (#13)
UPB-FILS-MA/upb-fils-ma.github.io: docs/lab/01.md

Reviewer: Rename it to `00.md`

    let x = 5;
    println!("The value of x is: {x}");
    x = 6;
    println!("The value of x is: {x}");
}
```

In this case, we will get a compilation error because we are trying to modify the value of `x` from `5` to `6`, but
`x` is immutable, so we cannot make this modification.

Although variables are immutable by default, you can make them **mutable** by adding `mut` in front of the
variable name. Adding `mut` also conveys intent to future readers of the code by indicating that other parts
of the code will modify the value of this variable.

```rust
fn main() {
    let mut x = 5;
    println!("The value of x is: {x}");
    x = 6;
    println!("The value of x is: {x}");
}
```

Now the value of `x` can become `6`.

### Constants

Like immutable variables, constants are values that are tied to a name and **are not allowed to change**,
but there are some differences between constants and variables.

First of all, you are not allowed to use `mut` with constants. Constants are not only immutable by default,
they are always immutable. You declare constants using the `const` keyword instead of the let keyword .

```rust
const THREE_HOURS_IN_SECONDS: u32 = 60 * 60 * 3;
```

> > To understand better, read [chapter 3](https://doc.rust-lang.org/book/ch03-00-common-programming-concepts.html) of the documentation.

Reviewer: Use a tip https://docusaurus.io/docs/markdown-features/admonitions

```rust
fn main() {
    let x = 2.0; // f64
    let y: f32 = 3.0; // f32
}
```

**Boolean** → Booleans are one byte in size. Boolean type in Rust is specified using bool.

```rust
let t = true;
let f: bool = false; // with explicit type annotation
```

**Character** → The Rust char type is the most primitive alphabetic type in the language.

```rust
let c = 'z';
let z: char = 'ℤ'; // with explicit type annotation
let heart_eyed_cat = '😻';
```

- ### Compound types

**Tuple** → A tuple is a general way of grouping a number of values ​​with a variety of types into a single compound type. Tuples have a **fixed** length  : once declared, their size cannot increase or decrease.

```rust
let tup: (i32, f64, u8) = (500, 6.4, 1);
```

**Array** → Unlike a tuple, each element in an array must have the **same type**. Unlike arrays in some other languages, Rust arrays have a **fixed** length.

```rust
let a = [1, 2, 3, 4, 5];
```

> > To understand better, read [chapter 3](https://doc.rust-lang.org/book/ch03-00-common-programming-concepts.html) of the documentation.

Reviewer: Use a tip https://docusaurus.io/docs/markdown-features/admonitions

### Functions

We define a function in Rust by entering fn followed by a function name and a set of parentheses. Curly braces tell the compiler where the function body begins and ends.

```rust
fn main() {
    println!("Hello, world!");
 
    another_function();
}
 
fn another_function() {
    println!("Another function.");
}
```
- ### Settings

We can define functions with parameters, which are special variables that are part of a function's signature. When a function has parameters, you can provide it with **concrete values** ​​for those parameters.

```rust
fn main() {
    another_function(5);
}
 
fn another_function(x: i32) {
    println!("The value of x is: {x}");
}
```

> > In function signatures you must declare the type of each parameter!

Reviewer: Use a tip https://docusaurus.io/docs/markdown-features/admonitions

- ### Declarations vs. expressions

Function bodies consist of a series of instructions optionally ending with an expression.

Declarations are **statements** that perform an action and do not return a value.

Expressions evaluate to a resulting value .

> > To understand better, read [chapter 3](https://doc.rust-lang.org/book/ch03-00-common-programming-concepts.html) of the documentation.

Reviewer: Use a tip https://docusaurus.io/docs/markdown-features/admonitions

- ### Functions with return values

Functions can return values ​​to the code that calls them. We don't name the return values, but we must declare their type after **an arrow (→)**. In Rust, the function's return value is synonymous with the value of **the final expression** in a function's body block. You can return earlier from a function by using the return keyword and specifying a value, but most functions implicitly return the last expression.

```rust
fn five() -> i32 {
    5
}
 
fn main() {
    let x = five();
    println!("The value of x is: {x}");// "The value of x is: 5"
}
```

> > To understand better, read [chapter 3](https://doc.rust-lang.org/book/ch03-00-common-programming-concepts.html) of the documentation.

Reviewer: Use a tip https://docusaurus.io/docs/markdown-features/admonitions

}
```
- ### Loop

The *loop* keyword tells Rust to run a block of code over and over forever or until you **explicitly** tell it to stop.

```rust
fn main() {
    loop {
        println!("again!");
    }
}
```
One use of a loop is to retry an operation that you know might fail, such as checking if a thread has finished its work. You may also need to pass the result of this operation out of the loop to the rest of your code. To do this, you can add the value you want to return after the break expression you use to stop the loop; this value will be returned out of the loop so you can use it:

```rust
fn main() {
    let mut counter = 0;
 
    let result = loop {
        counter += 1;
 
        if counter == 10 {
            break counter * 2;
        }
    };
 
    println!("The result is {result}");
}
```

- ### While
```rust
fn main() {
    let mut number = 3;
 
    while number != 0 {
        println!("{number}!");
 
        number -= 1;
    }
 
    println!("LIFTOFF!!!");
}
```

- ### For
```rust
fn main() {
    let a = [10, 20, 30, 40, 50];
 
    for element in a {
        println!("the value is: {element}");
    }
}
```

> > To understand better, read [chapter 3](https://doc.rust-lang.org/book/ch03-00-common-programming-concepts.html) of the documentation.

Reviewer: ```suggestion
> For a better understanding, please read [chapter 3](https://doc.rust-lang.org/book/ch03-00-common-programming-concepts.html) of the documentation.
```
Usa a tip https://docusaurus.io/docs/markdown-features/admonitions

```rust
    let mut s = String::new();
```
This line creates a new empty string called `s`, which we can then load data into.

**We can use the String::from** function or the **to_string()** function to create a string from a string literal:
```rust
    let s = String::from("initial contents");
```
```rust
let data = "initial contents";
 
    let s = data.to_string();
 
    // the method also works on a literal directly:
    let s = "initial contents".to_string();
```

- ### Adding to a stream

We can expand a string using the **push_str** method to add a string slice.
```rust
let mut s = String::from("foo");
s.push_str("bar");
```
**The push** method takes **a single character** as a parameter and adds it to the string.

```rust
    let mut s = String::from("lo");
    s.push('l');
```
- ### Iteration Methods on Strings

The best way to operate on pieces of strings is to be explicit about whether you want characters or bytes. For individual Unicode scalar values, use the **chars** method .
```rust
for c in "Зд".chars() {
    println!("{c}");
}
```
### Run the program
We must access the director where the main.rs file is located and execute the command:

```
cargo run
```

> ## Exercises

Reviewer: Should we add points to these?

Author: 2p given by default
1. - 1p
2. - 1p
3. - 1p
4. - 1p
5. - 2p
6. - 2p

Reviewer: Out of 10?

Author: I would grade every lab out of 10

---

# 01 - Rust

> ## LAB 00 - Rust for beginners

Reviewer: ```suggestion
## LAB 00 - Rust for beginners
```

> ## LAB 00 - Rust for beginners
> 

Reviewer: ```suggestion
```

> We will use the Rust programming language for the MA laboratories.

Reviewer: ```suggestion
We will use the [Rust](https://www.rust-lang.org/) programming language for the labs.
```

## Resources

> 1. The Rust Programming Language, Chapter 1, 2, 3 and 5 https://doc.rust-lang.org/book/#the-rustprogramming-language

Reviewer: ```suggestion
1. The Rust Programming Language, Chapters [1](), [2](), [3]() and [5]()
```

Add the link for each chapter.

> 2. Tour of Rust https://tourofrust.com/

Reviewer: ```suggestion
2. [Tour of Rust](https://tourofrust.com) step by step tutorial
```

## Basic programming language concepts for Rust

### Standard library

> By default, Rust has a set of elements defined in the standard library that it introduces into each program's

Reviewer: ```suggestion
By default, Rust has a set of elements defined in the standard library that it imports into each program's
```

> application framework. This set is called the Prelude, and you can see everything in it in the standard

Reviewer: ```suggestion
application framework. This set is called the *prelude*, and you can see everything in it in the standard
```

library [documentation](https://doc.rust-lang.org/std/prelude/index.html) .

> If a type you want to use is not in the prelude, you must bring that type into scope explicitly with a use

Reviewer: ```suggestion
If a type you want to use is not in the prelude, you must bring that type into scope explicitly with a `use`
```

> statement. Using the std::io library gives you a number of useful features, including the ability to accept

Reviewer: ```suggestion
statement. Using the `std::io` module gives you a number of useful features, including the ability to accept
```

user input.

> ``` 

Reviewer: ```suggestion
``` rust
```

use std::io; 
```

> ### The main function

Reviewer: ```suggestion
### The `main` function
```

> The main function is the entry point of our program.

Reviewer: ```suggestion
The `main` function is the entry point of our program.
```

> ``` 

Reviewer: ```suggestion
```rust 
```

fn main() {
    println!("Hello, world!");
}
```

> We use **println!()** to print messages on the screen.

Reviewer: ```suggestion
We use `println!` macro to print messages on the screen.
```

> To insert a placeholder in the println! method, use a **pair of braces {}** . We provide the variable name or

Reviewer: ```suggestion
To insert a placeholder in the `println!` method, use a *pair of braces {}* . We provide the variable name or
```

> expression to **replace** the provided placeholder outside the string.

Reviewer: ```suggestion
expression to replace the provided placeholder outside the string.
```

> ```

Reviewer: ```suggestion
```rust
```

fn main() {

    let name = "Mary";
    let age = 26;

>     println!("Hello, {}. You are {} years old", name, age);

Reviewer: ```suggestion
    println!("Hello, {}. You are {} years old", name, age);
    // if the replacements are only variable, one can use the inline version
    println!("Hello, {name}. You are {age} years old");
```

}
```

### Variables and mutability

> We use the let keyword to create a variable.

Reviewer: ```suggestion
We use the `let` keyword to create a variable.
```

> ```

Reviewer: ```suggestion
```rust
```

    let a = 5;
```

By default, in Rust, variables are **immutable** , meaning once a value is tied to a name, you cannot
change that value.

Example:

> ```

Reviewer: ```suggestion
```rust
```

fn main() {
    let x = 5;
    println!("The value of x is: {x}");
    x = 6;
    println!("The value of x is: {x}");
}
```

> In this case, we will get a compilation error because we are trying to modify the value of x from 5 to 6, but

Reviewer: ```suggestion
In this case, we will get a compilation error because we are trying to modify the value of `x` from `5` to `6`, but
```

x is immutable, so we cannot make this modification.

> Although variables are immutable by default, you can make them **mutable** by adding mut in front of the

Reviewer: ```suggestion
Although variables are immutable by default, you can make them **mutable** by adding `mut` in front of the
```

variable name. Adding mut also conveys intent to future readers of the code by indicating that other parts
of the code will modify the value of this variable.

> ```

Reviewer: ```suggestion
```rust
```

fn main() {
    let mut x = 5;
    println!("The value of x is: {x}");
    x = 6;
    println!("The value of x is: {x}");
}
```

> Now the value of x can become 6.

Reviewer: ```suggestion
Now the value of `x` can become `6`.
```

---

> # 01 - Rust

Reviewer: ```suggestion
# 00 - Rust
```
